### PR TITLE
reimplement bond / unbond

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

# Added
* payload.BondTx
* payload.UnbondTx

### CODE REVIEW

**Comment by gregdhill** [comment]
As BondTx reuses spec.TemplateAccount, we have the required fields of [IdentifyTx](https://github.com/hyperledger/burrow/blob/develop/docs/ADRs/draft/adr-2_identify-tx.md). In this scenario, a known validator should be able to submit an empty bond with the NodeAddress set.

File: cmd/burrow/commands/config_options.go [hunk 1]
```diff
@@ -10,102 +10,129 @@ import (
...
 		}),
 
 		externalAddressOpt: cmd.String(cli.StringOpt{
 			Name:   "x external-address",
 			Desc:   "An external address or host name provided with the port that this node will broadcast over gossip in order for other nodes to connect",
 			EnvVar: "BURROW_EXTERNAL_ADDRESS",
 		}),
+
+		configFileOpt: cmd.String(cli.StringOpt{
+			Name:   "c config",
```
Comment (silasdavis) [inline cmd/burrow/commands/config_options.go]: yes this makes sense

File: cmd/burrow/commands/start.go [hunk 2]
```diff
@@ -7,31 +7,19 @@ import (
...
-
 		configOpts := addConfigOptions(cmd)
 
 		cmd.Action = func() {
-			conf, err := obtainBurrowConfig(*configOpt, *genesisOpt)
-			if err != nil {
-				output.Fatalf("could not obtain config: %v", err)
-			}
-
-			err = configOpts.configure(conf)
```
Comment (silasdavis) [inline cmd/burrow/commands/start.go]: We are no longer configuring...

File: cmd/burrow/commands/configure.go [hunk 3]
```diff
@@ -313,42 +308,36 @@ func Configure(output Output) func(cmd *cli.Cmd) {
 				output.Fatalf("could not update burrow config: %v", err)
```
Comment (silasdavis) [inline cmd/burrow/commands/configure.go]: You should remove `configOpts.configure` on line 306 above - I think we can do it up front

File: cmd/burrow/commands/config_options.go [hunk 4]
```diff
@@ -10,102 +10,129 @@ import (
...
+
+		genesisFileOpt: cmd.String(cli.StringOpt{
+			Name:   "g genesis",
+			Desc:   "Use the specified genesis JSON file rather than a key in the main config, use - to read from STDIN",
+			EnvVar: "BURROW_GENESIS",
+		}),
+	}
+}
+
+func (opts *configOptions) initialize() (*config.BurrowConfig, error) {
```
Comment (silasdavis) [inline cmd/burrow/commands/config_options.go]: Rename this to `obtainBurrowConfig`

File: cmd/burrow/commands/config_options.go [hunk 5]
```diff
@@ -10,102 +10,129 @@ import (
...
+		return nil, fmt.Errorf("could not obtain config: %v", err)
+	}
+	err = opts.configure(conf)
+	if err != nil {
+		return nil, fmt.Errorf("could not configure config: %v", err)
 	}
+	return conf, nil
 }
 
 func (opts *configOptions) configure(conf *config.BurrowConfig) error {
```
Comment (silasdavis) [inline cmd/burrow/commands/config_options.go]: I think we can/should remove this entirely and make it a single shot in the function above

File: cmd/burrow/commands/configure.go [hunk 6]
```diff
@@ -38,24 +38,24 @@ func Configure(output Output) func(cmd *cli.Cmd) {
...
 			keys.DefaultKeysConfig().RemoteAddress))
 
-		keysDir := cmd.StringOpt("keysdir", "", "Directory where keys are stored")
+		keysDir := cmd.StringOpt("keys-dir", "", "Directory where keys are stored")
 
-		configOpt := cmd.StringOpt("c base-config", "", "Use the a specified burrow config file as a base")
+		configOpt := cmd.StringOpt("base-config", "", "Use the a specified burrow config file as a base")
 
-		genesisDocOpt := cmd.StringOpt("g genesis-doc", "", "GenesisDoc in JSON or TOML to embed in config")
+		genesisDocOpt := cmd.StringOpt("genesis-in", "", "GenesisDoc in JSON or TOML to embed in config")
```
Comment (silasdavis) [inline cmd/burrow/commands/configure.go]: We should not need this and `g genesis`

File: cmd/burrow/commands/config_options.go [hunk 7]
```diff
@@ -10,102 +10,129 @@ import (
...
 type configOptions struct {
-	validatorIndexOpt      *int
-	validatorAddressOpt    *string
-	validatorPassphraseOpt *string
-	validatorMonikerOpt    *string
-	externalAddressOpt     *string
+	configFileOpt        *string
+	genesisFileOpt       *string
+	accountIndexOpt      *int
+	accountAddressOpt    *string
```
Comment (silasdavis) [inline cmd/burrow/commands/config_options.go]: I'm not sure about the term 'account' for the validator address... we have many accounts, what's special about this one?

Comment (gregdhill) [reply cmd/burrow/commands/config_options.go]: How about `node`? I would just like to signify, particularly with bond/unbond, that the burrow process may not be running as a validator. However this is currently only true of accounts known in genesis.

Comment (gregdhill) [reply cmd/burrow/commands/config_options.go]: I've separated these review comments into #1113

**Comment by gregdhill** [comment]
# Considerations

## Delegated Bonding
Currently, there is no persistent mapping between input account(s) and a validator. Once bonded, only the validators respective account can update it's power or fully unbond. This is probably fine for the initial implementation, but we may want to consider more advanced mechanisms in the future whereby the respective accounts are credited based on their investment in a particular validator.

## Add / Remove Dynamics
Once an account with the right permission sends a BondTx, the target validator is immediately created and joins consensus after the following EndBlock. Similarly, an UnbondTx, signed by the submitting validator, is executed immediately. Hence I believe it is important to consider how we want to manage validators, from which this should be part of a larger discussion. For instance, upon completion of IdentifyTx (#960) we may want validators who leave the network to be automatically unbonded after a designated period of time.

File: execution/contexts/bond_context.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	// the account initiating the bond
+	account, err := ctx.StateWriter.GetAccount(ctx.tx.Input.Address)
+	if err != nil {
+		return err
+	}
+
+	// ensure pubKey of validator is set
+	val := ctx.tx.Validator
```
Comment (silasdavis) [inline execution/contexts/bond_context.go]: I don't think this is the correct type. We use `TemplateAccount` for `GovTx` because it can make arbitrary changes to an account. I think this should just be `PublicKey`

File: txs/payload/bond_tx.go [hunk 9]
```diff
@@ -5,12 +5,13 @@ import (
...
 	"github.com/hyperledger/burrow/crypto"
+	"github.com/hyperledger/burrow/genesis/spec"
 )
 
 func NewBondTx(pubkey crypto.PublicKey) (*BondTx, error) {
 	return &BondTx{
-		Inputs:   []*TxInput{},
-		UnbondTo: []*TxOutput{},
+		Input:     &TxInput{},
+		Validator: &spec.TemplateAccount{},
```
Comment (silasdavis) [inline txs/payload/bond_tx.go]: This should be a `PublicKey`

Perhaps the field should be `ValidatorPublicKey` for clarity (in JS at least)

File: execution/contexts/bond_context.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+		return fmt.Errorf("couldn't retrieve identity: %v", err)
+	}
+
+	// check if validator already exists
+	power, err := ctx.ValidatorSet.Power(*val.Address)
+	if err != nil {
+		return err
+	} else if power != nil && power.Cmp(big.NewInt(0)) == 1 && account.Address != *val.Address {
+		// we currently do not support delegated bonding
+		return fmt.Errorf("%s is already bonded", val.Address)
```
Comment (silasdavis) [inline execution/contexts/bond_context.go]: Seems like the error message here should be more like the comment. Also what is the non-positive power check for. Why do we continue if we are bonding 0 but input address does not match validator address?

File: execution/contexts/bond_context.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	priorPow, err := ctx.ValidatorSet.Power(*val.Address)
+	if err != nil {
+		return err
+	}
+	postPow := big.NewInt(0).Add(priorPow, power)
+	if !postPow.IsInt64() {
+		return fmt.Errorf("power supplied in update to validator power for %v does not fit into int64 and "+
+			"so is not supported by Tendermint", *val.Address)
+	}
```
Comment (silasdavis) [inline execution/contexts/bond_context.go]: None of the above checks should be necessary they are all handled by: https://github.com/hyperledger/burrow/blob/develop/acm/validator/bucket.go#L123-L132 and `Bucket` is part of the implementation of `Cache`.

So just try to do the `AddPower` later and check for an error

File: execution/contexts/bond_context.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+	valAcc, err := ctx.StateWriter.GetAccount(*val.Address)
+	if err != nil {
+		return err
+	} else if valAcc == nil {
+		valAcc = &acm.Account{
+			Address:     *val.Address,
+			PublicKey:   *val.PublicKey,
+			Sequence:    0,
+			Balance:     0,
+			Permissions: permission.NewAccountPermissions(permission.Bond),
```
Comment (silasdavis) [inline execution/contexts/bond_context.go]: Hm if this is an invariant we intend to rely on then we probably need to set it on the account we find if we find it.

Update; I do not think we should set this. Either validator == input and it must be already set or (in future) validator != input and the validator may not be able to self bond

File: execution/contexts/bond_context.go [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+		}
+		// pk must be known later to unbond
```
Comment (silasdavis) [inline execution/contexts/bond_context.go]: Let's not be lazy, public key is going to mean more to future generations

File: execution/contexts/bond_context.go [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	// create the account if it doesn't exist
+	valAcc, err := ctx.StateWriter.GetAccount(*val.Address)
+	if err != nil {
+		return err
+	} else if valAcc == nil {
+		valAcc = &acm.Account{
+			Address:     *val.Address,
+			PublicKey:   *val.PublicKey,
```
Comment (silasdavis) [inline execution/contexts/bond_context.go]: It's possible to valAcc to be in state but for us not to know the public key. e.g. if we send value to the account, only address will be known.

We collect public keys from input to successfully executed transactions here: https://github.com/hyperledger/burrow/blob/develop/execution/execution.go#L418 

In fact for the present implementation if our purpose is just to ensure we have the validator's public key, and currently the input must be the same address as validator bonding then `updateSignatories` will capture the public key after this tx goes through. So in this case we are probably fine.

However this is a special case, so we may want to be explicit for the case when validator != input

File: execution/contexts/bond_context.go [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	err = validator.AddPower(ctx.ValidatorSet, *val.PublicKey, power)
+	if err != nil {
+		return err
+	}
+
+	return ctx.StateWriter.UpdateAccount(account)
+}
+
+type UnbondContext struct {
```
Comment (silasdavis) [inline execution/contexts/bond_context.go]: I'm not sure we should depart from the 1 tx 1 context 1 file convention.

It's possible the bond and unbond can naturally collapse into a single transaction. It depends a bit on asymmetric between bonding and unbonding. But possibly we can govern investing and divesting by the magnitude and direction of a power change. i.e. unbonding and bonding periods may both make sense.

File: execution/contexts/bond_context.go [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+// Execute an UnbondTx to remove a validator
+func (ctx *UnbondContext) Execute(txe *exec.TxExecution, p payload.Payload) error {
+	var ok bool
+	ctx.tx, ok = p.(*payload.UnbondTx)
+	if !ok {
+		return fmt.Errorf("payload must be UnbondTx, but is: %v", txe.Envelope.Tx.Payload)
+	}
+
+	// the unbonding validator
+	sender, err := ctx.StateWriter.GetAccount(ctx.tx.Input.Address)
```
Comment (silasdavis) [inline execution/contexts/bond_context.go]: If we are going to tightly bind input ==  validator then we should do so in bond and remove the validator parameter altogether

File: execution/contexts/bond_context.go (hunk 10 above)
Comment (silasdavis) [reply execution/contexts/bond_context.go]: I guess it does nothing, but seems odd to continue.

File: execution/contexts/governance_context.go [hunk 17]
```diff
@@ -134,9 +115,36 @@ func (ctx *GovernanceContext) UpdateAccount(account *acm.Account, update *spec.T
 	return
 }
 
-func (ctx *GovernanceContext) MaybeGetPublicKey(address crypto.Address) (*crypto.PublicKey, error) {
+func GetIdentity(sw acmstate.ReaderWriter, account *spec.TemplateAccount) (err error) {
```
[the remark this answers is not in the document]
Comment (gregdhill) [reply execution/contexts/governance_context.go]: Agreed, with the simplification of bondTx this is not longer required.

File: execution/contexts/bond_context.go (hunk 8 above)
Comment (gregdhill) [reply execution/contexts/bond_context.go]: Updated

**Comment by gregdhill** [comment]
After a discussion with @silasdavis we've opted to more tightly bind the input account to the target validator, meaning less to go wrong. This also paves the way for economic primitives embedded in smart contracts for more advanced features such as delegated bonding.

File: execution/contexts/bond_context.go (hunk 15 above)
Comment (gregdhill) [reply execution/contexts/bond_context.go]: We had discussed combining bond / unbond into one transaction, but since TxInput amount is unsigned this would not be a trivial change. I've opted to keep these separate for now, but I have split into two context files as requested.

File: protobuf/payload.proto [hunk 18]
```diff
@@ -108,17 +108,18 @@ message BondTx {
     option (gogoproto.goproto_stringer) = false;
     option (gogoproto.goproto_getters) = false;
 
-    repeated TxInput Inputs = 1;
-    repeated TxOutput UnbondTo = 2;
+    // Account with bonding permission
+    TxInput Input = 1;
+    // PublicKey is retrieved from signature
```
Comment (silasdavis) [inline protobuf/payload.proto]: remove

File: acm/validator/bucket.go [hunk 19]
```diff
@@ -66,15 +66,17 @@ func (vc *Bucket) AlterPower(id crypto.PublicKey, power *big.Int) (*big.Int, err
 			"in a total power greater than that allowed by tendermint (%v): would make next total power: %v",
 			errHeader, id.GetAddress(), vc.Previous.GetPower(id.GetAddress()), power, maxTotalVotingPower, nextTotalPower)
 	}
+
 	// If we call vc.flow.ChangePower(id, absFlow) (below) will we induce a change in flow greater than the allowable
 	// flow we have left to spend?
-	if vc.Flow.Flow(id, absFlow).Cmp(allowableFlow) == 1 {
+	if vc.Flow.Flow(id, absFlow).Cmp(allowableFlow) == 1 && allowableFlow.Cmp(big.NewInt(0)) > 0 {
```
Comment (silasdavis) [inline acm/validator/bucket.go]: hm why is this needed? I think if allowableFlow is negative we have a bug, no?

Comment (gregdhill) [reply acm/validator/bucket.go]: `maxFlow` is `-1` if there are no validators

File: docs/ADRs/draft/adr-3_bonding-natives.md [hunk 20]
```diff
@@ -0,0 +1,40 @@
+---
+adr: 3
+title: Smart Contract Bonding Natives
+author: Silas Davis (@silasdavis), Gregory Hill (@gregdhill)
+discussions-to: https://chat.hyperledger.org/channel/burrow-contributors
+status: Draft
+type: Standards Track
```
[the remark this answers is not in the document]
Comment (gregdhill) [reply docs/ADRs/draft/adr-3_bonding-natives.md]: As in `Standards Track`? I'm not sure but it's what we used for the IdentifyTx proposal.

File: acm/validator/bucket.go (hunk 19 above)
Comment (silasdavis) [reply acm/validator/bucket.go]: As a bare minimum this needs a comment. But I don't particularly like handling it this way anyway. It's a special case so we should treat it as a special case rather than relying on this trick of the light.

In fact this is an important edge case of max flow...

Comment (silasdavis) [reply acm/validator/bucket.go]: We may want something like this, but I'm not sure yet, need to sleep on it: https://github.com/gregdhill/burrow/pull/2

**Comment by silasdavis** [comment]
I'm happy with all of this except the edge case we have idenitfied on Bucket - will try and come up with resolution to this next week. I think we should probably fix it now.
